0024-sub-state-returns.md

```diff
@@ -0,0 +1,153 @@
+only reports whether or not external state runs as a whole were successful and doesn't report individual changes.  
+I.E Return status of True if an entire Ansible playbook was executed successfully in an ansiblegate state
+
+We want to add the ability for Salt state returns to have a "sub_state_run" key, which gives
+these external state engines the ability to report the status of each of the "sub states" they
+ran independently, rather than reporting on the status of the entire group of sub states.
+
+The result is that state output will be very clear and verbose.
+Users will be able to see exactly which `idem state` passed/failed in an `idem SLS`,
+exactly which recipes passed/failed in a Chef cookbook, 
+and exactly which ansible plays passed/failed in a playbook 
+
+
+# Design
+[design]: #detailed-design
+
+Definitions
+-----------
+- External state engine: A Salt module that has the ability to execute a group of configurations (I.E Ansible, Chef, Idem)
+- Sub state run: A new key in the return of Salt states that allows definitions of external state engine run details
+
+
+How it is used
+--------------
+Some states can return multiple state runs from an external engine.
+State modules that extend tools like Ansible, Chef, and Idem can run multiple external
+"states" and then return their results individually in the "sub_state_run" portion of their return
+as long as their individual state runs are formatted like Salt states with low and high data.
+
+For example, the idem state module can execute multiple idem states
+via it's runtime and report the status of all those runs by attaching them to "sub_state_run" in it's state return.
+These sub_state_runs will be formatted and printed alongside other Salt states.
+
+Example
+-------
+
+```python
+state_return = {
+    "name": None,    # The parent state name
+    "result": None,  # The overall status of the external state engine run
+    "comment": None, # Comments on the overall external state engine run
+    "changes": {},   # An empty dictionary, each sub state run has it's own changes to report
+    "sub_state_run": [
+        {
+            "changes": {},       # A dictionary describing the changes made in the external state run
+            "result": None,      # The external state run name
+            "comment": None,     # Comment on the external state run
+            "duration": None,    # Optional, the duration in seconds of the external state run
+            "start_time": None,  # Optional, the timestamp of the external state run's start time
+            "low": {
+                "name": None,        # The name of the state from the external state run
+                "state": None,       # Name of the external state run
+                "__id__": None,      # ID of the external state run
+                "fun": None,         # The Function name from the external state run
+            },
+        }
+    ],
+}
```

Comment:
Is this a good way for the return data to be formatted?  I wanted to pass it through the state "low_data" function the same as regular states.  Any thoughts on the implementation here?

Reply:
> This talks a lot about the motivation and impact on `idem`, but what about the regular salt engine?
> What are the internal and observable changes to e.g. the ansible states, and would they require e.g. changes to peoples' returners?

By default, they won't see any change. It won't break, but nobody will see anything fancy - it will look exactly how it does today. *But* if the ansible states/modules are updated to actually return the sub state return data? 🎉  then they would show up. 

Basically this proposal is driven by `idem`, but can benefit anything that takes advantage of returning more structured data.

@Akm0d maybe it would also help to add a sample of the current output, as well as what the output would look like with this change applied - obviously using the default outputter, but perhaps JSON style as well.